win/rl/winrl.cc

```diff
@@ -470,6 +488,32 @@ NetHackRL::store_mapped_glyph(int ch, int color, int special, XCHAR_P x,
     specials_[offset] = special;
 }
 
+void
+NetHackRL::store_glyph_str(XCHAR_P x, XCHAR_P y, int glyph)
+{
+    // 1 <= x < cols, 0 <= y < rows (!)
+    size_t i = (x - 1) % (COLNO - 1);
+    size_t j = y % ROWNO;
+    size_t offset = j * (COLNO - 1) + i;
+
+    // see code in src/do_name.c:538 auto_describe
+    coord cc;
+    int sym = 0;
+    char tmpbuf[BUFSZ];
+    const char *firstmatch = "unknown";
+
+    cc.x = x;
+    cc.y = y;
+
+    if (do_screen_description(cc, TRUE, sym, tmpbuf, &firstmatch,
```

Comment:
(For tmpbuf. After allocating your string to some good length first... Or perhaps not.)

This `glyph_strs_[offset] = firstmatch;` creates a new `std::string` here every time which is likely more wasteful than any of the memcopies (which could be done w/o any allocations). But:

(1) I might be wrong.
(2) Even if I'm not, none of this is likely to matter.

Reply:
Also, to the extent that this is "expansive", it does still happen even if the observation is disabled, right? Because this `store_glyph_str` function is called nevetheless?

Reply:
I dont think you can pass `&glyph_strs[offset][0]` instead of `&firstmatch` but comment 1) has made me doubt myself, because I thought about this while writing.

`firstmatch` points to first letter of string literal `"unknown"`,  and we pass in the address of that pointer to change it so it now points to some other string `"description"'. (speaking of... where is this allocated? do i need to clean it up or is everything a string literal or static?).

`glyph_strs[offset][0]` this decays to a pointer of the first element in `glyphs_str[offset]`, and passing the address of that pointer does not guaranteer that the final string will be written in `glyph_strs_[offset][0]`. Besides if it did (!) how would it know how much space it is actually allowed to write in.

The unknown size of firstmatch is why i decided to go with `std::string`. We could do the whole thing as C string arrays, but wed have to be v v careful w null bytes (everywhere), as mentioned elsewhere.

Also yes it is expensive if not needed and we should probably wrap and exclude if possible. I figured that packing the whole dungeon arrays would be the slow part but maybe not

Reply:
This has now been changed to deal with C strings. We write NLE_GLYPH_STR_LENGTH of firstmatch into the correct array (and null terminate).

Reply:
Let's double check we have to do that. We might be able to use the argument you call `tmpbuf` :)

Reply:
When developing I printed tmpf to see what was in it anyway. The way the description function works is that it populates a list of objects (in the buffer) and then seems to chose one. Ie by the end `tmpbuf` ends up with:  

"a wall or floor of the room or a doorway (floor of the room)" and `firstmatch` becomes "floor of the room". I figure since we dont want to do our own parsing (probably), better just to take `firstmatch`. Happy to discuss/investigate more.

Reply:
Looking at the context in https://github.com/facebookresearch/nle/blob/master/src/pager.c#L1308, it appears all of what you call `tmpbuf` (it's called `out_str` there) is printed when you press `.` to select this thing. For the "hover" situation it only prints the "main part", although I'm not entirely sure where in the this happens. You can verify this by using `;` in the game.

You can get more information still by using `/` in the game. It will call `do_supplemental_info`. This will query some data file and show one more several nice quotes from books about the object in question.

Given that more info won't hurt _and_ it saves a memcopy, why not use the `out_str` argument? :)

Reply:
Discussed in person.
* `tmpbuf` is probably less useful than `firstmatch` since it has confusing information (stuff that matched but isnt whats being pointed to)
* `firstmatch` is clearly not pointing into `tmpbuff`, so we'll probably have to do a copy (cant just send in the array to write to.
* the big concern is the potential realloction of memory from the copy constructor. we use `assign` and `clear` to ensure we allcate once and just overwrite.